### PR TITLE
Add initializer for in-process Gradle executor

### PR DESCRIPTION
**Pull request by wolfs**
Avoiding strange static state caused by unit tests that causes flaky test failures.

The real culprit seems to be the reuse of the cache inside the `AsmBackedClassGenerator`: https://github.com/gradle/gradle/blob/d70090545b34a65edaf90843cd81fbaec9d152ac/subprojects/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java#L173-L179
If some test (e.g. `WorkerProcessIntegrationTest`) creates an `AsmBackedClassGenerator` with the `TestCrossBuildInMemoryCacheFactory` first, then all other tests will get a `TestCache` as the cache implementation. The `TestCache` is not thread-safe, and the `ConcurrentClassDecorationSpec` would break.

Making `TestCache` thread-safe fixes the immediate breakage of the `ConcurrentClassDecorationSpec`, but this is only a symptom and doesn't address the root cause. So we add the initializer for Global services instead so we don't get into the situation.

### CODE REVIEW

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecutorInitialization.groovy [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * For example, {@link org.gradle.internal.instantiation.generator.AsmBackedClassGenerator}.GENERATED_CLASSES_CACHES
+ * may have a test implementation instead of the real thing when a unit test runs first.
+ */
+@Issue("https://github.com/gradle/gradle-private/issues/3534")
+class InProcessGradleExecutorInitialization implements IGlobalExtension {
+    @Override
+    void start() {
+        if (GradleContextualExecuter.embedded ) {
+            AbstractGradleExecuter.GLOBAL_SERVICES.get(ObjectFactory)
+        }
```
Comment (lptr) [inline testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecutorInitialization.groovy:36]: Can we add some more comments about why this is happening only in embedded mode, and why this specific service is being queried?

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 2]
```diff
@@ -115,6 +116,7 @@ public abstract class AbstractGradleExecuter implements GradleExecuter, Resettab
         .build();
 
     private static final JvmVersionDetector JVM_VERSION_DETECTOR = GLOBAL_SERVICES.get(JvmVersionDetector.class);
+    private static final ObjectFactory objectFactory = GLOBAL_SERVICES.get(ObjectFactory.class);
```
Comment (lptr) [inline testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java:119]: Isn't this an unused field now? Can we clarify with a comment?

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecutorInitialization.groovy (hunk 1 above)
Comment (wolfs) [reply testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecutorInitialization.groovy:36]: Is the Javadoc on the class enough for describing why it only happens for embedded mode? I can add something here why we query `ObjectFactory`.

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java (hunk 2 above)
Comment (wolfs) [reply testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: This was a leftover, I removed it.

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecutorInitialization.groovy (hunk 1 above)
Comment (wolfs) [reply testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecutorInitialization.groovy]: I left some more comments.

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecutorInitialization.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Initializes services for the in-process Gradle executor.
+ * <p>
+ * If you don't initialize the services, some global state may end up with the wrong kind of values.
+ * For example, {@link org.gradle.internal.instantiation.generator.AsmBackedClassGenerator}.GENERATED_CLASSES_CACHES
+ * may have a test implementation instead of the real thing when a unit test runs first.
+ */
+@Issue("https://github.com/gradle/gradle-private/issues/3534")
+public class InProcessGradleExecutorInitialization implements IGlobalExtension {
```
Comment (mlopatkin) [inline testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecutorInitialization.java:31]: 💅 maybe reference this class in the InProcessGradleExecutor javadocs too? Global extensions are not something used often in our codebase (in fact, I've learned about them right now), and the extra visibility can save some time in the future for people figuring out how the executer is being initialized.

**Review by mlopatkin — APPROVED** [review]
Thanks! Comments really make it clear.

**Comment by wolfs** [comment]
@lptr We can't merge this since I didn't fix the build problems, yet. Hope to get to that this week.

File: testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecutorInitialization.java (hunk 3 above)
Comment (wolfs) [reply testing/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/InProcessGradleExecutorInitialization.java:31]: Done.

**Pull request merged**